dll/win32/shell32/wine/shellpath.c

```diff
@@ -490,23 +578,59 @@ BOOL WINAPI PathFileExistsAW (LPCVOID lpszPath)
  */
 BOOL WINAPI IsLFNDriveA(LPCSTR lpszPath)
 {
-    DWORD	fnlen;
-
-    if (!GetVolumeInformationA(lpszPath, NULL, 0, NULL, &fnlen, NULL, NULL, 0))
-	return FALSE;
-    return fnlen > 12;
+    WCHAR szBuffW[MAX_PATH], *pszW = NULL;
+    if (lpszPath)
+    {
+        SHAnsiToUnicode(lpszPath, szBuffW, _countof(szBuffW));
+        pszW = szBuffW;
+    }
+    return IsLFNDriveW(pszW);
 }
 
 /*************************************************************************
  * IsLFNDriveW		[SHELL32.42]
  */
 BOOL WINAPI IsLFNDriveW(LPCWSTR lpszPath)
 {
-    DWORD	fnlen;
+    DWORD cchMaxFileName, iDrive;
+    WCHAR szRoot[MAX_PATH];
 
-    if (!GetVolumeInformationW(lpszPath, NULL, 0, NULL, &fnlen, NULL, NULL, 0))
-	return FALSE;
-    return fnlen > 12;
+    if (lpszPath == NULL || lpszPath[0] == UNICODE_NULL)
+    {
+        szRoot[0] = 0;
+        GetWindowsDirectoryW(szRoot, _countof(szRoot));
+        lpszPath = szRoot;
+    }
+
+    if (PathIsUNCW(lpszPath))
+    {
+        StringCchCopyW(szRoot, _countof(szRoot), lpszPath);
+        PathStripToRootW(szRoot);
+
+        if (StrChrW(&szRoot[2], TEXT('\\')) == NULL)
+            return TRUE; /* LFN */
+
+        StringCchCatW(szRoot, _countof(szRoot), L"\\"); /* Add a backslash */
+    }
+    else /* Assuming absolute path... */
+    {
+        iDrive = ((lpszPath[0] - L'A') & 0x1F);
+        PathBuildRootW(szRoot, iDrive);
+
+        if (!IsRemovableDrive(iDrive))
+        {
+            /* FIXME: Cache correctly */
+        }
+    }
+
+#define MSDOS_8DOT3_LEN 12 /* MS-DOS 8.3 == length 12 */
+
+    /* I don't want to return FALSE if GetVolumeInformationW fails. See below */
+    cchMaxFileName = MSDOS_8DOT3_LEN + 1;
```

Comment:
Then it's better to change `cchMaxFileName` to this value when `GetVolumenInformationW` fails,
because now you depend on how `GetVolumenInformationW` treats arguments in a failure case.

Reply:
OK.